Avoid UB with self `pthread_join()`

mdavidsaver opened the pull request:
Attempt to address #480 by handling self-join in `epicsThreadMustJoin()` rather than depending on `pthread_join()`.

mdavidsaver committed:
avoid UB with self pthread_join()

exzombie commented on modules/libcom/src/osi/os/posix/osdThread.c:
```diff
@@ -682,32 +682,45 @@ void epicsThreadMustJoin(epicsThreadId id)
+    if(prev==0) {
+        /* join of unjoinable */
+        if(self==id) {
+            errlogPrintf(ERL_WARNING ": %s thread self-join of unjoinable\n", id->name);
+            return;
```
I think you are missing a call to `free_threadInfo()`

exzombie replied:
Or is this meant to be a complete no-op? The comment at the end of the function says "always decrement refcnt"; the "always" is confusing to me, and there are no docs.

mdavidsaver replied:
`prev==0` implies that `refcnt==1`.  For a self-join, this last reference is for the thread itself, and will be decremented when it actual exits.  This happens automatically through the use of a destructor function passed to `pthread_key_create()` when creating the thread local behind `epicsThreadGetIdSelf()`.

mdavidsaver replied:
I changed the wording of this comment to "release joinable reference".

mdavidsaver committed:
posix: epicsThreadCreateOpt() avoid leak of joinable on EPERM

On a system where RT priorities have been probed.
When creating a joinable thread, but still fails
with EPERM, the first epicsThreadOSD was leaked.

exzombie commented:
I was planning to open an issue against asyn because `asynPortDriver` causes a crash when EPICS is built with ThreadSanitizer ([this is the cause](https://github.com/epics-modules/asyn/blob/master/asyn/asynPortDriver/asynPortDriver.cpp#L934)). With this PR, the issue went away. It fixes more than just musl.

mdavidsaver commented:
I have added a fix for a memory leak I just noticed from `epicsThreadCreateOpt()` when `joinable` in the (I think) unlikely situation where initial SCHED_FIFO probing succeeded, but a later `pthread_create()` fails with `EPERM`.

ericonr commented on modules/libcom/src/osi/os/posix/osdThread.c:
```diff
@@ -689,32 +689,45 @@ void epicsThreadMustJoin(epicsThreadId id)
+        cantProceed(ERL_ERROR ": %s joins corrupt thread handle\n", self->name);
+
+    } else if(self==id) {
         /* Thread can't join itself (directly or indirectly)
          * so we detach instead.
          */
```
This comment needs to be updated, this block is for the thread joining itself directly.

ericonr reviewed (commented):
Doesn't this version lack the `pthread_detach` fallback when `pthread_join` fails with `EDEADLK` for valid calls?

mdavidsaver commented:
> Doesn't this version lack the `pthread_detach` fallback when `pthread_join` fails with `EDEADLK` for valid calls?

That it did.  Thank you for pointing this out.  Should now be correct.

anjohnson commented:
AI on @anjohnson : Examine closely, review.

ericonr reviewed (approved):
I can confirm that the test case now passed for me, on musl.
